### PR TITLE
Parse 'with' and resolve the mixins.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/compiler/parser.cc [hunk 1]
```diff
@@ -981,26 +981,32 @@ Declaration* Parser::parse_declaration(bool is_abstract) {
 }
 
 Class* Parser::parse_class_interface_monitor_or_mixin(bool is_abstract) {
+  auto at = [&](Symbol symbol) -> bool {
```
Comment (kasperl) [inline src/compiler/parser.cc:984]: I'd probably add a helper on Parser for this instead.

File: src/compiler/parser.cc [hunk 2]
```diff
@@ -1048,14 +1054,36 @@ Class* Parser::parse_class_interface_monitor_or_mixin(bool is_abstract) {
...
+
+    if (current_token() == Token::IDENTIFIER && !at(Symbols::implements) && !at(Symbols::with)) {
       super = parse_type(false);
     }
-    if (current_token() == Token::IDENTIFIER && current_token_data() == Symbols::implements) {
+    if (at(Symbols::with)) {
+      if (!requires_super) {
+        // "requires_super" indicates that there was an `extends`.
+        // TODO(florian): do we really want this requirement.
+        // The `class A with Foo` reads a bit worth than `class A extends Object with Foo`, but
```
Comment (kasperl) [inline src/compiler/parser.cc:1069]: worth? I don't understand the comment as is.

File: src/compiler/parser.cc [hunk 3]
```diff
@@ -1048,14 +1054,36 @@ Class* Parser::parse_class_interface_monitor_or_mixin(bool is_abstract) {
...
     name = parse_identifier();
     bool requires_super = false;
-    if (current_token() == Token::IDENTIFIER && current_token_data() == Symbols::extends) {
+    if (at(Symbols::extends)) {
       consume();
       requires_super = true;
     }
-    if (current_token() == Token::IDENTIFIER && current_token_data() != Symbols::implements) {
+
+    if (current_token() == Token::IDENTIFIER && !at(Symbols::implements) && !at(Symbols::with)) {
```
Comment (kasperl) [inline src/compiler/parser.cc:1062]: This feels like a poor use of `at`. I'd just check `current_token_data`.

File: src/compiler/parser.cc [hunk 4]
```diff
@@ -1048,14 +1054,36 @@ Class* Parser::parse_class_interface_monitor_or_mixin(bool is_abstract) {
...
+        // it still feels okish.
+        report_error("'with' requires an 'extends' clause");
+      }
+      if (super == null && requires_super) {
+        report_error("Missing super class");
+        // We reported an error. No need for a super class anymore.
+        requires_super = false;
+      }
+      consume();
+      while (current_token() == Token::IDENTIFIER && !at(Symbols::implements)) {
```
Comment (kasperl) [inline src/compiler/parser.cc:1079]: This feels like a poor use of `at`. I'd just check the `current_token_data` here too.

File: src/compiler/resolver.cc [hunk 5]
```diff
@@ -1228,11 +1244,26 @@ void Resolver::setup_inheritance(std::vector<Module*> modules, int core_module_i
...
 
+      auto ast_mixins = ast_class->mixins();
+      ListBuilder<ir::Class*> ir_mixins;
+      for (int i = 0; i < ast_mixins.length(); i++) {
+        auto ast_mixin = ast_mixins[i];
+        auto ir_mixin = resolve_class_interface_or_mixin(ast_mixin, scope, klass, false, true);
+        if (ir_mixin == null) {
+          report_error(ast_mixin, "Unresolved mixin");
+        } else if (!ir_mixin->is_mixin()) {
+          report_error(ast_mixin, "Not an mixin");
```
Comment (kasperl) [inline src/compiler/resolver.cc:1255]:
```suggestion
          report_error(ast_mixin, "Not a mixin");
```

File: tests/negative/gold/mixin-test.gold [hunk 6]
```diff
@@ -1,9 +1,36 @@
...
+tests/negative/mixin-test.toit:29:41: error: Unresolved mixin
+class ClassB extends Object with ClassA UnknownMixin some_method:
+                                        ^~~~~~~~~~~~
+tests/negative/mixin-test.toit:29:54: error: Unresolved mixin
+class ClassB extends Object with ClassA UnknownMixin some_method:
+                                                     ^~~~~~~~~~~
+tests/negative/mixin-test.toit:31:20: error: Super of a mixin must be a mixin
+mixin MixE extends Object with MixD:
+                   ^~~~~~
+tests/negative/mixin-test.toit:33:30: error: Not an mixin
```
Comment (kasperl) [inline tests/negative/gold/mixin-test.gold:31]:
```suggestion
tests/negative/mixin-test.toit:33:30: error: Not a mixin
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/compiler/parser.cc (hunk 1 above)
Comment (floitsch) [reply src/compiler/parser.cc:984]: Added `at_pseudo` (since this is basically a look for a pseudo keyword.

File: src/compiler/parser.cc (hunk 3 above)
Comment (floitsch) [reply src/compiler/parser.cc:1062]: done.

File: src/compiler/parser.cc (hunk 2 above)
Comment (floitsch) [reply src/compiler/parser.cc:1069]: Removed the TODO and comment. We can always change it, if we feel like it.
Comment was saying that `class A with Foo` isn't really English (compared to `class A extends B with Foo`), but that we would probably still be ok with it.
done.

File: src/compiler/parser.cc (hunk 4 above)
Comment (floitsch) [reply src/compiler/parser.cc:1079]: done.

File: src/compiler/resolver.cc (hunk 5 above)
Comment (floitsch) [reply src/compiler/resolver.cc:1255]: done.

File: tests/negative/gold/mixin-test.gold (hunk 6 above)
Comment (floitsch) [reply tests/negative/gold/mixin-test.gold:31]: done.

**Pull request merged**